src/navmesh.js

```diff
@@ -373,98 +373,113 @@ export class NavMesh {
+                extendLeft = true;
+                newPoint = portal.p2;
+            } else if (portal.p2.equals(lastRight)) {
+                extendLeft = true;
+                newPoint = portal.p1;
+            } else {
+                throw new Error("Invalid portal: Funnel cannot be extended on both sides at once.");
             }
 
-            edge1 = newEdge1;
-            edge2 = newEdge2;
-
-            // If we are at the end of the path,
-            // just add the last edge, which is from -> to
-            if (i === path.length - 2) {
-                const toEdge = new Edge(from, to);
-                let newToEdge = this._funnelEdge(edge1, edge2, toEdge);
-                if (newToEdge === edge1) {
-                    points.push(edge1.p2);
-                } else if (newToEdge === null) {
-                    points.push(edge2.p2);
-                }
+            // Extend funnel
+            this._extendFunnel(tail, left, right, extendLeft, newPoint);
+        }
+
+        // Close funnel to endpoint
+        this._extendFunnel(tail, left, right, true, to);
+        this._extendFunnel(tail, left, right, false, to);
+
+        return tail;
+    }
 
-                points.push(to);
+    _extendFunnel(tail, left, right, extendLeft, newPoint){
+        const apex = tail[tail.length - 1];
+        // We pretend to be in the `expandLeft` case here. Otherwise flip.
+        if (!extendLeft) {
+            [left, right] = [right, left];
+        }
+        // Determine angle of `apex`-`newPoint` relative to `apex`-`left[j]`
+        let j = 0;
+        while (j < left.length
+                && this._isInLeftRightOrder(apex, newPoint, left[j], !extendLeft)) {
+            j++;
+        }
+        // All points in `left` with index `< j` are right of `newPoint` and
+        // all points in `left` with index `>= j` are left of or at the same angle as `newPoint`. 
+        left.length = j; // Shrink funnel if `j < left.length`
+        left.push(newPoint);
+        if (j === 0) {
+            // If the funnel shrunk all the way on the left, it might collapse to the right.
+            // Determine how far it needs to collapse
+            let k = 0;
+            while (k < right.length
+                    && !this._isInLeftRightOrder(apex, newPoint, right[k], !extendLeft)) {
+                k++;
             }
+            // All points in `right` with index `< k` are left of or at the same angle as `newPoint` and
+            // all points in `right` with index `>= k` are right of `newPoint`. 
+            tail.push(...right.splice(0, k)); // Collapse funnel if `k > 0`
         }
+    }
 
-        return points;
+    /** 
+     * Are the points `p1` and `p2` in left-to-right order, viewed from `origin`?
+     * Checks for right-to-left order instead if `flip = true`.
+     * Returns `false` if the angles are equal.
+     */
+    _isInLeftRightOrder(origin, p1, p2, flip = false) {
```

Comment:
I find "left-to-right" confusing as description. Maybe it would be better to use clockwise and counter-clockwise, meaning the angle from origin starting from p1 and ending in p2. That would leave no possibility for wrong interpretation. Left and right does not make much sense in this context.

(I understand you're looking at that article for reference and nomenclature)

Reply:
With this I mean also renaming the function.